apisix/plugins/ocsp-stapling.lua

```diff
@@ -0,0 +1,253 @@
+        core.log.info("not ocsp resp cache found, fetch from ocsp responder")
+        ocsp_resp, err = fetch_ocsp_resp(der_cert_chain)
+        if ocsp_resp == nil then
+            return false, err
+        end
+        core.log.info("fetch ocsp resp ok, cache it")
+        ocsp_resp_cache:set(der_cert_chain, ocsp_resp, cache_ttl)
+    end
+
+    if not skip_verify then
+        local ok, err = ngx_ocsp.validate_ocsp_response(ocsp_resp, der_cert_chain)
+        if not ok then
+            return false, "failed to validate ocsp response: " .. err
+        end
+    end
+
+    -- set the OCSP stapling
+    local ok, err = ngx_ocsp.set_ocsp_status_resp(ocsp_resp)
+    if not ok then
+        return false, "failed to set ocsp status response: " .. err
+    end
+
+    return true
+end
+
+
+local original_set_cert_and_key
+local function set_cert_and_key(sni, value)
+    if value.gm then
+        -- should not run with gm plugin
+        core.log.info("gm plugin enabled, no need to run ocsp-stapling plugin")
+        return original_set_cert_and_key(sni, value)
+    end
+
+    if not value.ocsp_stapling then
+        core.log.info("no 'ocsp_stapling' field found, no need to run ocsp-stapling plugin")
+        return original_set_cert_and_key(sni, value)
+    end
+
+    if value.ocsp_stapling.enabled then
+        if not ngx.ctx.tls_ext_status_req then
+            core.log.info("no status request required, no need to send ocsp response")
+            return original_set_cert_and_key(sni, value)
+        end
+
+        local ok, err = set_pem_ssl_key(sni, value.cert, value.key)
+        if not ok then
+            return false, err
+        end
+        local fin_pem_cert = value.cert
+
+        -- multiple certificates support.
+        if value.certs then
+            for i = 1, #value.certs do
+                local cert = value.certs[i]
+                local key = value.keys[i]
+                ok, err = set_pem_ssl_key(sni, cert, key)
+                if not ok then
+                    return false, err
+                end
+                fin_pem_cert = cert
+            end
+        end
+
+        local ok, err = set_ocsp_resp(fin_pem_cert,
+                                      value.ocsp_stapling.skip_verify,
+                                      value.ocsp_stapling.cache_ttl)
+        if not ok then
+            core.log.error("no ocsp response send: ", err)
+        end
+
+        return true
+    end
+
+    return original_set_cert_and_key(sni, value)
+end
+
+
+function _M.init()
+    if core.schema.ssl.properties.gm ~= nil then
+        core.log.error("ocsp-stapling plugin should not run with gm plugin")
+    end
+
+    original_set_cert_and_key = radixtree_sni.set_cert_and_key
+    radixtree_sni.set_cert_and_key = set_cert_and_key
+
+    if core.schema.ssl.properties.ocsp_stapling ~= nil then
+        core.log.error("Field 'ocsp_stapling' is occupied")
+    end
+
+    core.schema.ssl.properties.ocsp_stapling = {
```

Comment:
Why do you insert a schema here?

Reply:
to provide options for controlling the details of ocsp stapling in each ssl configuration.